src/main/java/ch/njol/skript/sections/EffSecFireworkLaunch.java

```diff
@@ -0,0 +1,166 @@
+	"\tif damage cause is entity explosion:",
+	"\t\tif metadata value \"cancelDamage\" of event-projectile is true:",
+	"\t\t\tcancel event",
+	"",
+	"command /firework:",
+	"\ttrigger:",
+	"\t\tlaunch a firework with effects ball large coloured red at player:",
+	"\t\t\tset metadata value \"cancelDamage\" of event-firework to true"})
+@Since("2.4, INSERT VERSION (with section)")
+public class EffSecFireworkLaunch extends EffectSection {
+
+	public static class FireworkLaunchEvent extends Event {
+		private final Firework firework;
+
+		public FireworkLaunchEvent(Firework firework) {
+			this.firework = firework;
+		}
+
+		public Firework getFirework() {
+			return firework;
+		}
+
+		@Override
+		public HandlerList getHandlers() {
+			throw new IllegalStateException();
+		}
+	}
+
+	static {
+		Skript.registerSection(EffSecFireworkLaunch.class, "(launch|deploy) [[a] firework [with effect[s]]] %fireworkeffects% at %locations% [([with] (duration|power)|timed) %-number%]");
+		EventValues.registerEventValue(FireworkLaunchEvent.class, Firework.class, new Getter<Firework, FireworkLaunchEvent>() {
+			@Override
+			public Firework get(FireworkLaunchEvent fireworkLaunchEvent) {
+				return fireworkLaunchEvent.getFirework();
+			}
+		}, 0);
+	}
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<FireworkEffect> effects;
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<Location> locations;
+	@Nullable
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<Number> lifetime;
+
+	@Nullable
+	private Trigger trigger;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs,
+						int matchedPattern,
+						Kleenean isDelayed,
+						SkriptParser.ParseResult parseResult,
+						@Nullable SectionNode sectionNode,
+						@Nullable List<TriggerItem> triggerItems) {
+		effects = (Expression<FireworkEffect>) exprs[0];
+		locations = (Expression<Location>) exprs[1];
+		lifetime = (Expression<Number>) exprs[2];
+
+		if (sectionNode != null) {
+			trigger = loadCode(sectionNode, "fireworklaunch", FireworkLaunchEvent.class);
+		}
+
+		return true;
+	}
+
+	@Override
+	@Nullable
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	protected TriggerItem walk(Event e) {
+		Object localVars = Variables.copyLocalVariables(e);
+
+		Consumer<? extends Firework> consumer;
+		if (trigger != null) {
+			consumer = o -> {
+				FireworkLaunchEvent fireworkLaunchEvent = new FireworkLaunchEvent(o);
+				// Copy the local variables from the calling code to this section
+				Variables.setLocalVariables(fireworkLaunchEvent, localVars);
+				trigger.execute(fireworkLaunchEvent);
+			};
+		} else {
+			consumer = null;
+		}
+
+		Number power = lifetime.getSingle(e);
+		if (power == null)
+			power = 1;
+		for (Location location : locations.getArray(e)) {
+			Firework firework = location.getWorld().spawn(location, Firework.class, (Consumer) consumer);
+			FireworkMeta meta = firework.getFireworkMeta();
```

Comment:
Is it possible to make the meta changes within the consumer?
If so, you should supply a consumer that makes the meta changes, and then runs the trigger consumer.

something like
```java
(firework) -> {
			FireworkMeta meta = firework.getFireworkMeta();
			meta.addEffects(effects.getArray(e));
			meta.setPower(power.intValue());
			firework.setFireworkMeta(meta);
			consumer.accept(firework);
}
```

Reply:
I just tested it, it doesn't work.